pkg/controller/route/routecontroller.go

```diff
@@ -33,79 +43,217 @@ type RouteController struct {
+	key, err := controller.KeyFunc(obj)
 	if err != nil {
-		return fmt.Errorf("error listing routes: %v", err)
+		glog.Errorf("Couldn't get key for object %+v: %v", obj, err)
+		return
 	}
-	// TODO (cjcullen): use pkg/controller/framework.NewInformer to watch this
-	// and reduce the number of lists needed.
-	nodeList, err := rc.kubeClient.Nodes().List(api.ListOptions{})
+	rc.queue.Add(key)
+}
+
+func (rc *RouteController) syncNodeRoute(key string) error {
+	startTime := time.Now()
+	defer func() {
+		glog.V(4).Infof("Finished syncing route %q (%v)", key, time.Now().Sub(startTime))
+	}()
+
+	var node *api.Node
+	obj, exists, err := rc.nodeStore.Store.GetByKey(key)
 	if err != nil {
-		return fmt.Errorf("error listing nodes: %v", err)
+		glog.Infof("Unable to retrieve node %v from store: %v", key, err)
+		rc.queue.Add(key)
+		return err
+	}
+	if exists {
+		node = obj.(*api.Node)
 	}
-	return rc.reconcile(nodeList.Items, routeList)
+
+	var r *cloudprovider.Route
+	routes, err := rc.routes.ListRoutes(rc.clusterName)
+	if err != nil {
+		return fmt.Errorf("error listing routes: %v", err)
+	}
+	// Check if we have a route for this node
+	// TODO(mqliang): we should compute this once and dynamically update it using Watch, not constantly re-compute.
+	for _, route := range routes {
+		if route.TargetInstance != key {
+			continue
+		}
+		r = route
+	}
+
+	var shouldDelete bool
+	var shouldCreate bool
+
+	// if we have a route but the node has been deleted, remove the route
+	if node == nil && r != nil {
+		glog.Infof("Node %v has been deleted, remove it's route", key)
+		shouldDelete = true
+	}
+	// if we have no route for node, create one
+	if node != nil && r == nil {
+		glog.Infof("No route for Node %v, create one", key)
+		shouldCreate = true
+	}
+	// check if the route for this node has the correct CIDR, if not, remove the old route and create a new one
+	if node != nil && r != nil && r.DestinationCIDR != node.Spec.PodCIDR {
+		glog.Infof("The route for %v has the incorrect CIDR, remove it and create a new one", key)
+		shouldDelete = true
+		shouldCreate = true
+	}
+
+	if shouldDelete {
+		if err := rc.routes.DeleteRoute(rc.clusterName, r); err != nil {
+			glog.Errorf("Could not delete route %s %s: %v", r.Name, r.DestinationCIDR, err)
+		}
+	}
+
+	if shouldCreate {
+		route := &cloudprovider.Route{
+			TargetInstance:  node.Name,
+			DestinationCIDR: node.Spec.PodCIDR,
+		}
+		nameHint := string(node.UID)
+		if err := rc.routes.CreateRoute(rc.clusterName, nameHint, route); err != nil {
+			glog.Errorf("Could not create route %s %s: %v", nameHint, route.DestinationCIDR, err)
+		}
+	}
+
+	return nil
 }
 
-func (rc *RouteController) reconcile(nodes []api.Node, routes []*cloudprovider.Route) error {
+// checkLeftoverRoutes lists all currently existing routes and detect routes
+// that exist with no corresponding nodes; these routes need to be deleted.
+// We only need to do this once on startup, because in steady-state these are
```

Comment:
Consider such a case:
1. Add a node A, RouteController create a route RouteA for this node
2. We fix several bugs in ControllerManager and want upgrade it, so we stop ControllerManager for several minute.
3. A user delete the Node A
4. ControllerManager restart, and since Node A has been deleted, Informer will never observe a Event about Node A. So, there remains a route with no corresponding node.